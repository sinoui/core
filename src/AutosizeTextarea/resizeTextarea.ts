/* eslint-disable no-param-reassign */
function getSizeData(element: HTMLElement) {
  const computedStyle = getComputedStyle(element);
  const getSizeValue = (key: keyof CSSStyleDeclaration) =>
    parseInt(computedStyle[key], 10) || 0;
  const { boxSizing } = computedStyle;
  const padding = getSizeValue('paddingTop') + getSizeValue('paddingBottom');
  const border = getSizeValue('borderTop') + getSizeValue('borderBottom');

  return { boxSizing, padding, border };
}

function getContentHeight(shadow: HTMLTextAreaElement, content = 'x') {
  shadow.value = content;
  return shadow.scrollHeight;
}

/**
 * 根据内容重新计算文本域的尺寸
 *
 * @param textarea 需要变更尺寸的文本域
 * @param shadow 用来计算内容尺寸的文本域
 * @param minRows 最小行数
 * @param maxRows 最大行数
 */
export default function resizeTextarea(
  textarea: HTMLTextAreaElement,
  shadow: HTMLTextAreaElement,
  minRows?: number,
  maxRows?: number,
) {
  const { boxSizing, padding, border } = getSizeData(textarea);
  const contentHeight =
    getContentHeight(shadow, textarea.value || textarea.placeholder || 'x') -
    padding;
  const singleRowHeight = getContentHeight(shadow) - padding;
  let textareaHeight = contentHeight;
  if (minRows) {
    textareaHeight = Math.max(singleRowHeight * minRows, textareaHeight);
  }
  if (maxRows) {
    textareaHeight = Math.min(singleRowHeight * maxRows, textareaHeight);
  }

  const newHeight =
    textareaHeight + (boxSizing === 'border-box' ? padding + border : 0);

  textarea.style.height = `${newHeight}px`;
  textarea.style.overflow =
    contentHeight - textareaHeight >= 1 ? 'auto' : 'hidden';
}
