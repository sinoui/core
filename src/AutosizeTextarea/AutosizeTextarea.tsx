import React, { useRef, useEffect, useCallback } from 'react';
import { debounce } from '@sinoui/utils';
import useMultiRefs from '../utils/useMultiRefs';
import resizeTextarea from './resizeTextarea';

interface AutosizeTextareaProps
  extends React.ComponentPropsWithoutRef<'textarea'> {
  /**
   * 指定最小展示行数
   */
  minRows?: number;
  /**
   * 指定最大展示行数
   */
  maxRows?: number;
}

const shadowStyle: React.CSSProperties = {
  position: 'absolute',
  top: 0,
  left: 0,
  maxHeight: 0,
  minHeight: 0,
  height: 0,
  visibility: 'hidden',
  pointerEvents: 'none',
  transform: 'translateZ(0)',
  overflow: 'hidden',
};

const useEnhancedEffect =
  typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;

/**
 * 自动计算大小的多行文本域
 */
const AutosizeTextarea = React.forwardRef<
  HTMLTextAreaElement,
  AutosizeTextareaProps
>((props, ref) => {
  const { minRows, maxRows, ...rest } = props;
  const textareaRef = useRef<HTMLTextAreaElement>(null);
  const shadowRef = useRef<HTMLTextAreaElement>(null);
  const textareaRefHandler = useMultiRefs(ref, textareaRef);

  const resizeTextareaCallback = useCallback(() => {
    const shadow = shadowRef.current;
    const textarea = textareaRef.current;
    if (!shadow || !textarea) {
      return;
    }

    resizeTextarea(textarea, shadow, minRows, maxRows);
  }, [maxRows, minRows]);

  useEnhancedEffect(resizeTextareaCallback);

  useEffect(() => {
    const resize = debounce(resizeTextareaCallback);
    window.addEventListener('resize', resize, false);
    return () => {
      resize.cancel();
      window.removeEventListener('resize', resize, false);
    };
  }, [resizeTextareaCallback]);

  return (
    <>
      <textarea ref={textareaRefHandler} rows={minRows} {...rest} />
      <textarea
        className={rest.className}
        ref={shadowRef}
        tabIndex={-1}
        aria-hidden="true"
        style={{ ...rest.style, ...shadowStyle }}
      />
    </>
  );
});

export default AutosizeTextarea;
