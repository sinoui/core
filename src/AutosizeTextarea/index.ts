import AutosizeTextarea from './AutosizeTextarea';

export default AutosizeTextarea;
