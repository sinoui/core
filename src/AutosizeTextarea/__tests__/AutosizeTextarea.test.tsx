import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import AutosizeTextarea from '../AutosizeTextarea';

it('渲染文本域', () => {
  const tree = renderer.create(<AutosizeTextarea />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('给文本域添加样式', () => {
  const tree = renderer
    .create(
      <AutosizeTextarea
        style={{
          boxSizing: 'border-box',
        }}
      />,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('在非受控模式自动调整高度', () => {
  const changeCallback = jest.fn();
  const { getByTestId } = render(
    <AutosizeTextarea data-testid="textarea" onChange={changeCallback} />,
  );

  const textarea = getByTestId('textarea');

  fireEvent.change(textarea, { target: { value: '123' } });

  expect(changeCallback).toBeCalled();
});
