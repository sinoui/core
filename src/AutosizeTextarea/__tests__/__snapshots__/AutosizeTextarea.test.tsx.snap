// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染文本域 1`] = `
Array [
  <textarea
    onChange={[Function]}
  />,
  <textarea
    aria-hidden="true"
    style={
      Object {
        "height": 0,
        "left": 0,
        "maxHeight": 0,
        "minHeight": 0,
        "overflow": "hidden",
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
        "transform": "translateZ(0)",
        "visibility": "hidden",
      }
    }
    tabIndex={-1}
  />,
]
`;

exports[`给文本域添加样式 1`] = `
Array [
  <textarea
    onChange={[Function]}
    style={
      Object {
        "boxSizing": "border-box",
      }
    }
  />,
  <textarea
    aria-hidden="true"
    style={
      Object {
        "boxSizing": "border-box",
        "height": 0,
        "left": 0,
        "maxHeight": 0,
        "minHeight": 0,
        "overflow": "hidden",
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
        "transform": "translateZ(0)",
        "visibility": "hidden",
      }
    }
    tabIndex={-1}
  />,
]
`;
